### ISSUE TITLE
Rule proposal: `prefer-t-throws`

### ISSUE THREAD

**Issue by gajus** [avajs/eslint-plugin-ava#156]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2482%20Rewarded-%237E24E3.svg" />](https://oss.issuehunt.io/r/avajs/eslint-plugin-ava/issues/156)

I was reviewing some codes using ava and found this pattern repeating often:

```js
try {
  await request(requestOptions);
} catch (error) {
  t.true(error.statusCode === 500);
}
```

This can be written instead as:

```js
const error = await t.throws(request(requestOptions));

t.true(error.statusCode === 500);
```

I think the latter should be preferred.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $82.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://oss.issuehunt.io/u/issuehunt) ($80.00)
- [<img src='https://avatars.githubusercontent.com/u/170270?v=4' alt='sindresorhus' width=24 height=24> sindresorhus](https://oss.issuehunt.io/u/sindresorhus) ($2.00)

### Submitted pull Requests
- [#360 fix(prefer-t-throws): detect try/catch asserting on caught error](https://oss.issuehunt.io/r/avajs/eslint-plugin-ava/pull/360)
- [#835 Add `request` and `options` to `beforeError` hook state](https://oss.issuehunt.io/r/sindresorhus/ky/pull/835)
---

### Tips

- Checkout the [Issuehunt explorer](https://oss.issuehunt.io/r/avajs/eslint-plugin-ava/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://oss.issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by jfmengels** [comment]
Sounds like a good idea to me :+1:. Not sure when/why try/catch makes sense in tests when t.throws is available.

The second example is better, but FYI, using `t.fail()` would improve your first example.
```js
try {
  await request(requestOptions);
  t.fail();
} catch (error) {
  t.true(error.statusCode === 500);
}
```

**Comment by gajus** [comment]
> The second example is better, but FYI, using t.fail() would improve your first example.

Another idea for a lint rule. :-)

**Comment by sindresorhus** [comment]
👍 

I think `prefer-t-throws` would be a better name. We don't want to prevent all usage of try/catch. Thoughts?

**Comment by jfmengels** [comment]
Agreed

**Comment by gajus** [comment]
> I think prefer-t-throws would be a better name. We don't want to prevent all usage of try/catch. Thoughts?

What would be a valid use case for `try..catch` that cannot translate to `t.throws`?

**Comment by jfmengels** [comment]
I think that `t.throws()` handles every case you'd use a try catch for, and even the `Promise#catch()`.

**Comment by sindresorhus** [comment]
@gajus Maybe something that might throw, but that you don't care about in the test, so you'd like to silence it. I'm sure there are other cases we can't think of too. But the main point is that the intent is clearer with `prefer-t-throws` than `no-try-catch`; We want to recommend using `t.throws()`, not arbitrarily prevent try/catch.

**Comment by vdemedes** [comment]
The immediate valid use case for `try..catch` that came up in my mind is error handling in Koa:

```js
app.use(function * (next) {
  try {
    yield* next;
  } catch (err) {
    // handle error
  }  
});
```

**Comment by vdemedes** [comment]
> We don't want to prevent all usage of try/catch. 

Definitely agree with this.

**Comment by IssuehuntBot** [comment]
@issuehunt has funded $80.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/49722492/issues/156) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/49722492/issues/156).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.
